src/common.c

```diff
@@ -1187,20 +1188,13 @@ int ofi_pollfds_wait(struct ofi_pollfds *pfds, void **contexts,
 		fastlock_release(&pfds->lock);
 
 		/* Index 0 is the internal signaling fd, skip it */
-		for (i = pfds->index; i < pfds->nfds && found < max_contexts; i++) {
```

Comment:
I thought about whether to keep this feature or not, updating index in cleanup.  The general purpose behind this abstraction is to emulate epoll semantics.  And I don't know that epoll provides a similar optimization.  If the app provides a large enough array, this optimization should do very little.  All events will be reported out, and the app will simply iterate over them.  There's only the possibility of starvation if the input array is too small.  Do we know if epoll handles this case differently?

Because the release is targeted for today (external schedule pressure), I can keep the index to minimize the changes.

Reply:
The fairness index is set to the last index which reported an event.  If the app passes in max_contexts = 1, it's possible that it will only get events from a single fd.  A better fairness fix is needed if we want to keep this.

Reply:
The index also changes during the loops.  So instead of loop 1 being from X to nfds, and loop 2 being from 0 to X, loop 2 can actually overlap with with loop 1.  That is, loop 2 can go from 0 to X+Y, resulting in the same fd being reported twice.  That could also be part of the issue.  IMO, the best option at this point is to remove the buggy code and go with a simpler solution.  We'll need to figure out fairness later in a way that's correct.